activerecord/lib/active_record/associations/preloader.rb

```diff
@@ -107,26 +107,27 @@ def preload(records, associations, preload_scope = nil)
 
       private
 
-      def preloaders_on(association, records, scope)
+      def preloaders_on(association, records, scope, options = {})
         case association
         when Hash
-          preloaders_for_hash(association, records, scope)
+          preloaders_for_hash(association, records, scope, options)
         when Symbol
-          preloaders_for_one(association, records, scope)
+          preloaders_for_one(association, records, scope, options)
         when String
-          preloaders_for_one(association.to_sym, records, scope)
+          preloaders_for_one(association.to_sym, records, scope, options)
         else
           raise ArgumentError, "#{association.inspect} was not recognised for preload"
         end
       end
 
-      def preloaders_for_hash(association, records, scope)
+      def preloaders_for_hash(association, records, scope, options = {})
         association.flat_map { |parent, child|
           loaders = preloaders_for_one parent, records, scope
+          polymorphic = options[:polymorphic] || loaders.any?{ |l| l.reflection.polymorphic? }
```

Comment:
This line doesn't work when `l` is a `NullPreloader` - perhaps `NullPreloader` should expose the `owners` and `reflection`, as `AlreadyLoaded` does - after all they are part of the public interface of a preloader.